### PR TITLE
Add missing algorithm macro checks

### PR DESCRIPTION
**Pull request by ben-dav-lytle**
Ran into compile errors when disabling/enabling certain algorithms.

Signed-off-by: Ben Lytle <ben.dav.lytle@hpe.com>

### CODE REVIEW

**Comment by stefanberger** [comment]
> Ran into compile errors when disabling/enabling certain algorithms.

Can you describe how you ran into these issues?

File: src/tpm2/SymmetricTest.h [hunk 1]
```diff
@@ -84,7 +84,7 @@ const SYMMETRIC_TEST_VECTOR   c_symTestValues[NUM_SYMS + 1] = {
       dataOut_AES128_CFB, dataOut_AES128_ECB}},
 #endif
 #if ALG_AES && AES_192
-    {TPM_ALG_AES_, 192, key_AES192, 16, sizeof(dataIn_AES192), dataIn_AES192,
+    {TPM_ALG_AES, 192, key_AES192, 16, sizeof(dataIn_AES192), dataIn_AES192,
```
Comment (stefanberger) [inline src/tpm2/SymmetricTest.h:87]: Do you need AES_192? Iirc I left this in there on purpose since libtpms doesn't support it in its default config.

**Comment by stefanberger** [comment]
> Ran into compile errors when disabling/enabling certain algorithms.

Ok giot i

**Comment by stefanberger** [comment]
I did not mean to close the PR.

File: src/tpm2/crypto/CryptSym.h [hunk 2]
```diff
@@ -102,7 +102,10 @@
 //#define     TDES    DES[3]			/* libtpms commented */
 typedef union tpmCryptKeySchedule_t {
     FOR_EACH_SYM_WITHOUT_TDES(KEY_SCHEDULE)	/* libtpms changed from FOR_EACH_SYM */
+
+#if ALG_TDES
```
Comment (stefanberger) [inline src/tpm2/crypto/CryptSym.h:106]: Please append `// libtpms added`

File: src/tpm2/crypto/CryptSym.h [hunk 3]
```diff
@@ -102,7 +102,10 @@
...
     tpmKeyScheduleTDES  tdes[3];		/* libtpms added */
+#endif
```
Comment (stefanberger) [inline src/tpm2/crypto/CryptSym.h:108]: Please append `// libtpms added`

File: src/tpm2/crypto/openssl/CryptEccSignature.c [hunk 4]
```diff
@@ -104,6 +104,7 @@ EcdsaDigest(
 /* e) n is the order of the curve */
 /* Error Returns Meaning */
 /* TPM_RC_NO_RESULT the result of the operation was zero or r (mod n) is zero */
+#if ALG_ECSCHNORR
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/CryptEccSignature.c:107]: Please append `// libtpms added`

File: src/tpm2/crypto/openssl/CryptEccSignature.c [hunk 5]
```diff
@@ -130,6 +131,7 @@ BnSchnorrSign(
     BnDiv(NULL, bnS, bnT1, bnN);
     return (BnEqualZero(bnS)) ? TPM_RC_NO_RESULT : TPM_RC_SUCCESS;
 }
+#endif // ALG_ECSHNORR
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/CryptEccSignature.c:134]: Please append `// libtpms added`

**Comment by ben-dav-lytle** [comment]
> > Ran into compile errors when disabling/enabling certain algorithms.
> 
> Can you describe how you ran into these issues?

Sorry, should've included that in PR description.

Specifically, I disabled `ALG_TDES` and `ALG_ECSHNORR` and got "undefined" compile errors where the new `#if` checks are.
I enabled AES_192, to test, and that was the only place that failed compilation.

I am testing different security profiles as far as supporting/not supporting specific algorithms; like FIPS 140-2 and 140-3, CNSA 1.0 and CNSA 2.0

**Comment by stefanberger** [comment]
> I am testing different security profiles as far as supporting/not supporting specific algorithms; like FIPS 140-2 and 140-3, CNSA 1.0 and CNSA 2.0

These are TPM security profiles? The problem for libtpms is basically that we cannot really have compile-time options for security profiles for a TPM 2 but algorithms have to be made runtime-invisibel/inaccessible. See my PR #343 for those types of profiles.  The state of the TPM 2 and possibly downgrading requirements are a major challenge...

**Comment by ben-dav-lytle** [comment]
> > I am testing different security profiles as far as supporting/not supporting specific algorithms; like FIPS 140-2 and 140-3, CNSA 1.0 and CNSA 2.0
> 
> These are TPM security profiles? The problem for libtpms is basically that we cannot really have compile-time options for security profiles for a TPM 2 but algorithms have to be made runtime-invisibel/inaccessible. See my PR #343 for those types of profiles. The state of the TPM 2 and possibly downgrading requirements are a major challenge...

Not TPM security profiles, but just conforming to FIPS or CNSA, as far as what algorithms are supported.

Yes, so similar to what you propose with #343; I was trying to take it one step further as completely compiling out the algorithms, but state migration would obviously not work. Is #343 still in testing?

I see you mention the profiles could only configure algorithms and not commands. Future plan for commands?

**Comment by stefanberger** [comment]
> Yes, so similar to what you propose with #343; I was trying to take it one step further as completely compiling out the algorithms, but state migration would obviously not work. Is #343 still in testing?

I am deferring #343. Any later version of libtpms, whatever may be added to it, needs to be able to write the state of a previous version of libtpms driven by the applied profile. This can become a major headache including not being able to fix data in the state (which did happen before).

Compiling out algorithms is not really a choice because there will only be one version of libtpms in distros and it basically should enable all algorithms (and commands) and make algorithms and commands inaccessible if a certain profile is chosen. So it for example implements TDES but TDES is inaccessible & invisible (TPM2_GetCapability) because the profile doesn't support it.

> I see you mention the profiles could only configure algorithms and not commands. Future plan for commands?

Not currently...

**Comment by ben-dav-lytle** [comment]
> Compiling out algorithms is not really a choice because there will only be one version of libtpms in distros and it basically should enable all algorithms (and commands) and make algorithms and commands inaccessible if a certain profile is chosen. So it for example implements TDES but TDES is inaccessible & invisible (TPM2_GetCapability) because the profile doesn't support it.

Understood.
I'm testing configure options. If no option for `--with-only-algs` is supplied, all algorithms are available (or compiled) as normal:
```
./autogen.sh --with-openssl --with-tpm2 --without-tpm1 --with-only-algs=fips140-3 --prefix=/usr && make -j`nproc`
```
Other values would be `fips140-2` or `cnsa1` or `cnsa2`. It can kind of convolute algorithm inclusion in  `TpmProfile.h`, though.

**Comment by stefanberger** [comment]
> Understood. I'm testing configure options. If no option for `--with-only-algs` is supplied, all algorithms are available (or compiled) as normal:
> 
> ```
> ./autogen.sh --with-openssl --with-tpm2 --without-tpm1 --with-only-algs=fips140-3 --prefix=/usr && make -j`nproc`
> ```
> 
> Other values would be `fips140-2` or `cnsa1` or `cnsa2`. It can kind of convolute algorithm inclusion in `TpmProfile.h`, though.

I understand but IMO there will only be one version of libtpms that has as many algorithms enabled as there are possible and the disablement happens via profiles ...

**Comment by stefanberger** [comment]
Cab you prefix your commit messages with 'tpm2:'?

**Comment by stefanberger** [comment]
... and squash the patches into one...

**Comment by ben-dav-lytle** [comment]
> Can you prefix your commit messages with 'tpm2:'?
... and squash the patches into one...

Done!

**Comment by stefanberger** [comment]
Thanks.

**Pull request merged**